curp/tests/it/common/curp_group.rs

```diff
@@ -61,141 +64,169 @@ pub struct CurpGroup {
-                        handle,
-                        trigger,
-                    },
                 )
-            })
-            .collect();
+                .await,
+            );
+            let handle = tokio::spawn(Self::run(server, listener, l));
+            let curp_node = CurpNode {
+                id,
+                addr,
+                exe_rx,
+                as_rx,
+                role_change_arc,
+                handle,
+                trigger,
+            };
+            nodes.insert(curp_node.id, curp_node);
+        }
 
         tokio::time::sleep(Duration::from_millis(300)).await;
         debug!("successfully start group");
         Self {
             nodes,
-            storage_path,
+            storage_path: None,
         }
     }
 
+    async fn gen_listeners(keys: impl Iterator<Item = &String>) -> HashMap<String, TcpListener> {
+        join_all(
+            keys.cloned()
+                .map(|name| async { (name, TcpListener::bind("0.0.0.0:0").await.unwrap()) }),
+        )
+        .await
+        .into_iter()
+        .collect()
+    }
+
+    fn listeners_to_all_members_addrs(
+        listeners: &HashMap<String, TcpListener>,
+    ) -> HashMap<String, Vec<String>> {
+        listeners
+            .iter()
+            .map(|(name, listener)| {
+                (
+                    name.clone(),
+                    vec![listener.local_addr().unwrap().to_string()],
+                )
+            })
+            .collect()
+    }
+
+    fn get_snapshot_allocator_from_cfg(config: &CurpConfig) -> Box<dyn SnapshotAllocator> {
+        match config.engine_cfg {
+            EngineConfig::Memory => {
+                Box::<MemorySnapshotAllocator>::default() as Box<dyn SnapshotAllocator>
+            }
+            EngineConfig::RocksDB(_) => {
+                Box::<RocksSnapshotAllocator>::default() as Box<dyn SnapshotAllocator>
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    async fn run(
+        server: Arc<Rpc<TestCommand, TestRoleChange>>,
+        listener: TcpListener,
+        mut shutdown_listener: shutdown::Listener,
+    ) -> Result<(), tonic::transport::Error> {
+        tonic::transport::Server::builder()
+            .add_service(ProtocolServer::from_arc(Arc::clone(&server)))
+            .add_service(InnerProtocolServer::from_arc(server))
+            .serve_with_incoming_shutdown(
+                TcpListenerStream::new(listener),
+                shutdown_listener.wait_self_shutdown(),
+            )
+            .await
+    }
+
     pub async fn run_node(
         &mut self,
         listener: TcpListener,
         name: String,
         cluster_info: Arc<ClusterInfo>,
+    ) {
+        self.run_node_with_config(
+            listener,
+            name,
+            cluster_info,
+            Arc::new(CurpConfig::default()),
+            EngineConfig::default(),
+        )
+        .await
+    }
+
+    pub async fn run_node_with_config(
```

Comment:
The method `run_node_with_config` is only used in the `run_node` method, therefore it dosen't need to be specified as pub. BTW, it seems we only use default config in our test. Shall we inline this method into `run_node`?

Reply:
Sometimes we need to test configs that are not enabled by default, such as TLS, and we need to pass in config from the outside to test these things